substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/SecurityServicesFeature.java

```diff
@@ -342,11 +344,27 @@ public ProviderList cleanUnregisteredProviders(ProviderList providerList) {
             filteredProviders.removeIf(this::shouldRemoveProvider);
             if (filteredProviderList == null || !filteredProviderList.providers().equals(filteredProviders)) {
                 filteredProviderList = ProviderList.newList(filteredProviders.toArray(new Provider[0]));
+                removedProviders = new ArrayList<>(providerList.providers());
+                removedProviders.removeIf(provider -> !shouldRemoveProvider(provider));
```

Comment:
Maybe it would be worth guarding this like this:
```suggestion
                if (Options.TraceSecurityServices.getValue()) {
                    removedProviders = new ArrayList<>(providerList.providers());
                    removedProviders.removeIf(provider -> !shouldRemoveProvider(provider));
                }
```

Reply:
While I've initially done this without a guard for consistency with other tracing, I agree with you - it may be better to only compute this list only when needed to avoid overhead during non-debugging builds